sound/soc/sof/debug-dsp-ops.c

```diff
@@ -8,6 +8,97 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+/*
+ * set dsp power state op by writing the power state.
+ * ex: echo set_power_state,D3 > dsp_test_op
+ */
+static int sof_dsp_ops_set_power_state(struct snd_sof_dev *sdev, char *state)
+{
+	/* only D3 supported for now */
+	if (strcmp(state, "D3")) {
+		dev_err(sdev->dev, "Unsupported state %s\n", state);
+		return -EINVAL;
+	}
+
+	/* power off the DSP */
+	if (sdev->dsp_power_state.state == SOF_DSP_PM_D0) {
+		const struct sof_ipc_pm_ops *pm_ops = sof_ipc_get_ops(sdev, pm);
+		pm_message_t pm_state;
+		int ret;
+
+		pm_state.event = SOF_DSP_PM_D3;
+
+		/* suspend DMA trace */
+		sof_fw_trace_suspend(sdev, pm_state);
+
+		/* notify DSP of upcoming power down */
+		if (pm_ops && pm_ops->ctx_save) {
+			ret = pm_ops->ctx_save(sdev);
+			if (ret < 0)
+				return ret;
+		}
+
+		ret = snd_sof_dsp_runtime_suspend(sdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to power off DSP\n");
+			return ret;
+		}
+
+		sdev->enabled_cores_mask = 0;
+		sof_set_fw_state(sdev, SOF_FW_BOOT_NOT_STARTED);
+	}
+
+	return 0;
+}
+
+static int sof_dsp_ops_boot_firmware(struct snd_sof_dev *sdev)
+{
+	const char *fw_filename;
+	int ret;
+
+	/* power off the DSP */
+	ret = sof_dsp_ops_set_power_state(sdev, "D3");
+	if (ret < 0)
+		return ret;
+
+	if (sdev->basefw.fw)
+		snd_sof_fw_unload(sdev);
+
+	/* free trace */
+	sof_fw_trace_free(sdev);
+
+	/* power up the DSP */
+	ret = snd_sof_dsp_runtime_resume(sdev);
```

Comment:
Yikes. power up the DSP and dsp_runtime_resume doesn't seem like the same thing at all to me.